File: libmuscle/python/libmuscle/snapshot_manager.py
import logging
from pathlib import Path
from typing import cast, List, Optional

from ymmsl import Reference, Operator, Settings

from libmuscle.communicator import Message
from libmuscle.port_manager import PortManager
from libmuscle.mmp_client import MMPClient
from libmuscle.snapshot import MsgPackSnapshot, Snapshot, SnapshotMetadata

_logger = logging.getLogger(__name__)

_MAX_FILE_EXISTS_CHECK = 10000

# error text for save_snapshot when msg = None
_NO_MESSAGE_PROVIDED = (
        'Invalid message provided to `{}`. Please create a Message object to'
        ' store the state of the instance in a snapshot.')


class SnapshotManager:
    """Manages information on snapshots for the Instance

    Implements the saving and loading of snapshots in the checkpointing API.
    """

    def __init__(
            self, instance_id: Reference, manager: MMPClient,
            port_manager: PortManager) -> None:
        """Create a new snapshot manager

        Args:
            instance_id: The id of this instance.
            manager: The client used to submit data to the manager.
            port_manager: The port manager belonging to this instance.
        """
        self._instance_id = instance_id
        # replace identifier[i] by identifier-i to use in snapshot file name
        # using a dash (-) because that is not allowed in Identifiers
        self._safe_id = str(instance_id).replace("[", "-").replace("]", "")
        self._port_manager = port_manager
        self._manager = manager

        self._resume_from_snapshot: Optional[Snapshot] = None
        self.resume_overlay: Optional[Settings] = None
        self._next_snapshot_num = 1

    def prepare_resume(
            self, resume_snapshot: Optional[Path],
            snapshot_directory: Optional[Path]) -> Optional[float]:
        """Apply checkpoint info received from the manager.

        If there is a snapshot to resume from, this loads it and does
        any resume work that libmuscle should do, including restoring
        message counts and storing the resumed-from snapshot again as
        our first snapshot.

        Args:
            resume_snapshot: Snapshot to resume from (or None if not
                resuming)
            snapshot_directory: directory to save snapshots in

        Returns:
            Time at which the initial snapshot was saved, if resuming.
        """
        result: Optional[float] = None
        self._snapshot_directory = snapshot_directory or Path.cwd()
        if resume_snapshot is not None:
            snapshot = self.load_snapshot_from_file(resume_snapshot)

            if snapshot.message is not None:
                # snapshot.message is None for implicit snapshots
                self._resume_from_snapshot = snapshot
                result = snapshot.message.timestamp
            self.resume_overlay = snapshot.settings_overlay

            self._port_manager.restore_message_counts(snapshot.port_message_counts)
            # Store a copy of the snapshot in the current run directory
            path = self.__store_snapshot(snapshot)
            metadata = SnapshotMetadata.from_snapshot(snapshot, str(path))
            self._manager.submit_snapshot_metadata(metadata)

        return result

    def resuming_from_intermediate(self) -> bool:
        """Check whether we have an intermediate snapshot.

        Doesn't say whether we should resume now, just that we were
        given an intermediate snapshot to resume from by the manager.
        """
        return (
                self._resume_from_snapshot is not None and
                not self._resume_from_snapshot.is_final_snapshot)

    def resuming_from_final(self) -> bool:
        """Check whether we have a final snapshot.

        Doesn't say whether we should resume now, just that we were
        given an intermediate snapshot to resume from by the manager.
        """
        return (
                self._resume_from_snapshot is not None and
                self._resume_from_snapshot.is_final_snapshot)

    def load_snapshot(self) -> Message:
        """Get the Message to resume from.
        """
        snapshot = cast(Snapshot, self._resume_from_snapshot)
        return cast(Message, snapshot.message)

    def save_snapshot(
            self, msg: Optional[Message], final: bool,
            triggers: List[str], wallclock_time: float,
            f_init_max_timestamp: Optional[float],
            settings_overlay: Settings
            ) -> float:
        """Save a (final) snapshot.

        Args:
            msg: Message object representing the snapshot.
            final: True iff called from save_final_snapshot.
            triggers: Description of checkpoints that triggered this.
            wallclock_time: Wallclock time when saving.
            f_init_max_timestamp: Timestamp for final snapshots.
            settings_overlay: Current settings overlay.

        Returns:
            Simulation time at which the snapshot was made.
        """
        port_message_counts = self._port_manager.get_message_counts()
        if final:
            # Decrease F_INIT port counts by one: F_INIT messages are already
            # pre-received, but not yet processed by the user code. Therefore,
            # the snapshot state should treat these as not-received.
            all_ports = self._port_manager.list_ports()
            ports = all_ports.get(Operator.F_INIT, [])
            if self._port_manager.settings_in_connected():
                ports.append('muscle_settings_in')
            for port_name in ports:
                new_counts = [i - 1 for i in port_message_counts[port_name]]
                port_message_counts[port_name] = new_counts

        snapshot = MsgPackSnapshot(
                triggers, wallclock_time, port_message_counts, final, msg,
                settings_overlay)

        path = self.__store_snapshot(snapshot)
        metadata = SnapshotMetadata.from_snapshot(snapshot, str(path))
        self._manager.submit_snapshot_metadata(metadata)

        timestamp = msg.timestamp if msg is not None else float('-inf')
        if final and f_init_max_timestamp is not None:
            # For final snapshots f_init_max_snapshot is the reference time (see
            # should_save_final_snapshot).
            timestamp = f_init_max_timestamp
        return timestamp

    @staticmethod
    def load_snapshot_from_file(snapshot_location: Path) -> Snapshot:
        """Load a previously stored snapshot from the filesystem.

        Args:
            snapshot_location: path where the snapshot is stored.
        """
        _logger.debug(f'Loading snapshot from {snapshot_location}')
        if not snapshot_location.is_file():
            raise RuntimeError(f'Unable to load snapshot: {snapshot_location}'
                               ' is not a file. Please ensure this path exists'
                               ' and can be read.')

        # TODO: encapsulate I/O errors?
        with snapshot_location.open('rb') as snapshot_file:
            version = snapshot_file.read(1)
            data = snapshot_file.read()

            if version == MsgPackSnapshot.SNAPSHOT_VERSION_BYTE:
                return MsgPackSnapshot.from_bytes(data)
            raise RuntimeError('Unable to load snapshot from'
                               f' {snapshot_location}: unknown version of'
                               ' snapshot file. Was the file saved with a'
                               ' different version of libmuscle or'
                               ' edited?')

    def __store_snapshot(self, snapshot: Snapshot) -> Path:
        """Store a snapshot on the filesystem.

        Args:
            snapshot: Snapshot to store.

        Returns:
            Path where the snapshot is stored.
        """
        _logger.debug(f'Saving snapshot to {self._snapshot_directory}')
        for _ in range(_MAX_FILE_EXISTS_CHECK):
            # Expectation is that muscle_snapshot_directory is empty initially
            # and we succeed in the first loop. Still wrapping in a for-loop
            # such that an existing filename doesn't immediately raise an error
            fname = f"{self._safe_id}_{self._next_snapshot_num}.pack"
            fpath = self._snapshot_directory / fname
            self._next_snapshot_num += 1
            if not fpath.exists():
                break
        else:
            raise RuntimeError('Could not find an available filename for'
                               f' storing the next snapshot: {fpath} already'
                               ' exists.')
        # Opening with mode 'x' since a file with the same name may be created
        # in the small window between checking above and opening here. It is
        # better to fail with an error than to overwrite an existing file.
        with fpath.open('xb') as snapshot_file:
            snapshot_file.write(snapshot.SNAPSHOT_VERSION_BYTE)
            snapshot_file.write(snapshot.to_bytes())
        return fpath

File: libmuscle/cpp/src/libmuscle/snapshot_manager.cpp
#include <libmuscle/snapshot_manager.hpp>

#include <libmuscle/data.hpp>

#include <algorithm>
#include <cmath>
#include <fstream>
#include <sstream>

// Note: using POSIX for filesystem calls
// Could be upgraded to std::filesystem when targeting C++17 or later
#include <errno.h>
#include <fcntl.h>
#include <limits.h>
#include <string.h>
#include <unistd.h>

#define MAX_FILE_EXISTS_CHECK 10000

using namespace std::string_literals;

namespace libmuscle { namespace _MUSCLE_IMPL_NS {

SnapshotManager::SnapshotManager(
            ymmsl::Reference const & instance_id,
            MMPClient & manager,
            PortManager & port_manager,
            Logger & logger)
        : instance_id_(instance_id)
        , manager_(manager)
        , port_manager_(port_manager)
        , logger_(logger)
        , resume_from_snapshot_()
        , resume_overlay_()
        , next_snapshot_num_(1)
        , snapshot_directory_()
{
    // replace identifier[i] by identifier-i to use in snapshot file name
    // using a dash (-) because that is not allowed in Identifiers
    safe_id_ = static_cast<std::string>(instance_id_);
    std::replace(safe_id_.begin(), safe_id_.end(), '[', '-');
    safe_id_.erase(
            std::remove(safe_id_.begin(), safe_id_.end(), ']'),
            safe_id_.end());
}

namespace {

std::string string_error(int err_num) {
    char err_buf[1024];
#ifdef _GNU_SOURCE
    // g++ defines this by default, which uses the gnu extension for strerror_r
    return strerror_r(err_num, err_buf, 1024);
#else  // _GNU_SOURCE
    // POSIX complient strerror_r
    if (strerror_r(err_num, err_buf, 1024) == 0)
        return err_buf;
    return std::to_string(err_num);
#endif  // _GNU_SOURCE
}

// Adapted from https://stackoverflow.com/a/2203853
std::string get_working_path()
{
    char temp [ PATH_MAX ];

    if ( getcwd(temp, PATH_MAX) != 0)
        return std::string ( temp );

    throw std::runtime_error(
            "Error retrieving current working directory: " + string_error(errno));
}

}

Optional<double> SnapshotManager::prepare_resume(
        Optional<std::string> const & resume_snapshot,
        Optional<std::string> const & snapshot_directory) {
    Optional<double> result;

    if (snapshot_directory.is_set()) {
        snapshot_directory_ = snapshot_directory.get();
    } else {
        // store snapshots in current working directory
        snapshot_directory_ = get_working_path();
    }

    if (resume_snapshot.is_set()) {
        auto snapshot = load_snapshot_from_file(resume_snapshot.get());

        if (snapshot.message.is_set()) {
            // snapshot.message is not set for implicit snapshots
            resume_from_snapshot_ = snapshot;
            result = snapshot.message.get().timestamp();
        }
        resume_overlay_ = snapshot.settings_overlay;

        port_manager_.restore_message_counts(snapshot.port_message_counts);
        // Store a copy of the snapshot in the current run directory
        auto path = store_snapshot_(snapshot);
        auto metadata = SnapshotMetadata::from_snapshot(snapshot, path);
        manager_.submit_snapshot_metadata(metadata);
    }

    return result;
}

bool SnapshotManager::resuming_from_intermediate() {
    return (resume_from_snapshot_.is_set() &&
            !resume_from_snapshot_.get().is_final_snapshot);
}

bool SnapshotManager::resuming_from_final() {
    return (resume_from_snapshot_.is_set() &&
            resume_from_snapshot_.get().is_final_snapshot);
}

Message SnapshotManager::load_snapshot() {
    if (!resume_from_snapshot_.is_set())
        throw std::runtime_error("Error: no snapshot to load.");
    return resume_from_snapshot_.get().message.get();
}

double SnapshotManager::save_snapshot(
        Optional<Message> message, bool is_final,
        std::vector<std::string> const & triggers, double wallclock_time,
        Optional<double> f_init_max_timestamp,
        ::ymmsl::Settings settings_overlay) {
    auto port_message_counts = port_manager_.get_message_counts();

    if (is_final) {
        // Decrease F_INIT port counts by one: F_INIT messages are already
        // pre-received, but not yet processed by the user code. Therefore,
        // the snapshot state should treat these as not-received.
        auto all_ports = port_manager_.list_ports();
        auto ports = all_ports.find(::ymmsl::Operator::F_INIT);
        if (ports != all_ports.end()) {
            for (auto const & port_name : ports->second) {
                for (auto & count : port_message_counts[port_name]) {
                    --count;
                }
            }
        }
        if (port_manager_.settings_in_connected()) {
            for (auto & count : port_message_counts["muscle_settings_in"]) {
                --count;
            }
        }
    }

    Snapshot snapshot(
            triggers, wallclock_time, port_message_counts, is_final, message,
            settings_overlay);

    auto path = store_snapshot_(snapshot);
    auto metadata = SnapshotMetadata::from_snapshot(snapshot, path);
    manager_.submit_snapshot_metadata(metadata);

    double timestamp = message.is_set() ? message.get().timestamp() : -INFINITY;
    if (is_final && f_init_max_timestamp.is_set()) {
        // For final snapshots f_init_max_snapshot is the reference time (see
        // should_save_final_snapshot).
        timestamp = f_init_max_timestamp.get();
    }
    return timestamp;
}

Snapshot SnapshotManager::load_snapshot_from_file(
        std::string const & snapshot_location) {
    logger_.debug("Loading snapshot from " + snapshot_location);

    std::ifstream snapshot_file(snapshot_location, std::ios::binary);
    if (!snapshot_file.good()) {
        throw std::runtime_error(
                "Unable to load snapshot: " + snapshot_location +
                " cannot be opened. Please ensure this path exists and can be read.");
    }

    // get the size of the file
    snapshot_file.seekg(0, std::ios::end);
    std::size_t fsize = snapshot_file.tellg();
    snapshot_file.seekg(0, std::ios::beg);

    if (fsize < 1) {
        throw std::runtime_error(
                "Unable to load snapshot: " + snapshot_location + " is an empty file.");
    }

    char version;
    std::vector<char> data(fsize - 1);
    snapshot_file.read(&version, 1);
    snapshot_file.read(data.data(), fsize - 1);
    if (!snapshot_file.good()) {
        throw std::runtime_error(
                "Unable to load snapshot file " + snapshot_location +
                ". I/O error while reading file.");
    }

    switch (static_cast<Snapshot::VersionByte>(version)) {
        case Snapshot::VersionByte::MESSAGEPACK:
            return Snapshot::from_bytes(data);
        default:
            throw std::runtime_error(
                    "Unable to load snapshot file " + snapshot_location +
                    ": unknown version of snapshot file. Was the file saved with a"
                    " different version of libmuscle or edited?");
    }
}

std::string SnapshotManager::store_snapshot_(Snapshot const & snapshot) {
    logger_.debug("Saving snapshot to " + snapshot_directory_);
    std::ofstream snapshot_file;
    std::string fpath;
    for (int i=0; i<MAX_FILE_EXISTS_CHECK; ++i) {
        // Expectation is that muscle_snapshot_directory is empty initially
        // and we succeed in the first loop. Still wrapping in a for-loop
        // such that an existing filename doesn't immediately raise an error
        std::ostringstream fname;
        fname << snapshot_directory_ << "/";
        fname << safe_id_ << "_" << next_snapshot_num_ << ".pack";
        fpath = fname.str();
        ++next_snapshot_num_;

        // TODO: use std::ios::noreplace (from C++23) in fstream.open instead of this
        int fd = ::open(fpath.c_str(), O_WRONLY | O_CREAT | O_EXCL, 0777);
        if (fd < 0) {
            if (errno == EEXIST)
                continue;  // file already exists: retry next available
            throw std::runtime_error(string_error(errno));
        }
        ::close(fd);

        snapshot_file.open(fpath, std::ios::binary);
        if (snapshot_file.good()) {
            break;
        } else {
            snapshot_file.close();
        }
    }
    if (!snapshot_file.good()) {
        throw std::runtime_error(
                "Could not find an available filename for storing the next"
                " snapshot: " + fpath + " cannot be opened for writing.");
    }
    snapshot_file << static_cast<char>(Snapshot::VersionByte::MESSAGEPACK);
    auto data = snapshot.to_bytes();
    snapshot_file.write(data.data(), data.size());
    snapshot_file.close();
    if (!snapshot_file.good()) {
        throw std::runtime_error("I/O error while writing snapshot to file " + fpath);
    }
    return fpath;
}

} }
